### PR TITLE
Add example for dynamically mapped tasks

### PR DESCRIPTION
**Pull request by potiuk**
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragement file, named `{pr_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/example_dags/example_mapped_task.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+# specific language governing permissions and limitations
+# under the License.
+
+from typing import Iterable
+
+import pendulum
+
+from airflow import DAG, XComArg
+from airflow.decorators import task
+from airflow.operators.bash_operator import BashOperator
```
Comment (josh-fell) [inline airflow/example_dags/example_mapped_task.py]:
```suggestion
from airflow.operators.bash import BashOperator
```

Comment (josh-fell) [inline airflow/example_dags/example_mapped_task.py]:
```suggestion
from airflow.operators.bash import BashOperator
```

**Review by josh-fell — COMMENTED** [review]
Otherwise LGTM

File: airflow/example_dags/example_mapped_task.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    tags=['example'],
+) as dag:
+
+    @task
+    def make_list():
+        return [1, 2, 3]
+
+    @task
+    def add_one(x: int, multiplier: int = 1):
+        return (x + 1) * 2
```
Comment (ephraimbuddy) [inline airflow/example_dags/example_mapped_task.py]:
```suggestion
        return (x + multiplier) * 2
```

Comment (jedcunningham) [reply airflow/example_dags/example_mapped_task.py]: Good call, but shouldn't it be:
```suggestion
        return (x + 1) * multiplier
```

Comment (potiuk) [reply airflow/example_dags/example_mapped_task.py]: Right. I focused on getting it complex and try out all the features of dymamic tasks, not to be correct :)

File: airflow/example_dags/example_mapped_task.py (hunk 1 above)
Comment (potiuk) [reply airflow/example_dags/example_mapped_task.py]: Aha. Right.

File: airflow/example_dags/example_mapped_task.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    def echo_task(value: int):
+        return f"echo {value}"
+
+    added_values = add_one.partial(multiplier=2).expand(x=make_list())
+    second = add_one.expand(x=make_list())
+    third = add_one.expand(x=make_list(), multiplier=[2, 3, 4])
+    bash_input = echo_task.expand(value=third)
+    sum_it(third)
+
+    classic = BashOperator.partial(task_id="final").expand(bash_command=bash_input)
```
Comment (potiuk) [inline airflow/example_dags/example_mapped_task.py]:
```suggestion
    classic = BashOperator.partial(task_id="classic").expand(bash_command=bash_input)
```

**Review by josh-fell — APPROVED** [review]

**Comment by potiuk** [comment]
Seems like bitnami chart is failing more and more ...

**Review by pingzh — APPROVED** [review]
nice, thanks

**Review by ephraimbuddy — APPROVED** [review]

**Comment by ephraimbuddy** [comment]
Serialization error...

**Comment by potiuk** [comment]
This is strange :)

**Comment by uranusjr** [comment]
Looks like the _test_ is wrong. MappedOperator instances are deserialised into MappedOperator, not SerializedBaseOperator. The test should be updated to

```python
if serialized_task.is_mapped:
    assert isinstance(serialized_task, MappedOperator)
    assert isinstance(task, MappedOperator)
else:
    assert isinstance(serialized_task, SerializedBaseOperator)
    assert isinstance(task, BaseOperator)
assert not isinstance(task, SerializedBaseOperator)
```

**Comment by potiuk** [comment]
Thanks @uranusjr !

**Comment by potiuk** [comment]
Hey @uranusjr - actually I needed to do quite a number of other changes to make the tests pass. I am not sure however if some of those changes are how they should - be some of them were based rather on intuition than facts as the current hierarchy of operators and how it interacts with Serialization is well,, convoluted to say the least (most of it obviously due to history of adding MappedOperator on top of existing BaseOperator).

* I believe we had qute a bit of inconsistency in how we treated hash/eq of operators for BasedOperator and MappedOperator. I  think I fixed it - though when I tried to move it level up (to AbstractOperator) I had some strange "MapperOperator" is not hashable even if it derived from AbstractOperator - but likely some magic of serializing/deserializing which I don't fully understand caused it. 

* operators_extra_links in baseOperator were )_ rather than [] and it cause comparision with serialized version wrong (this is the one I have most reservations about - likely it shoudl be changed elsewhere

* XComArg() is recreated when serializing and since is passed as value in a dictionary, the field comparision flagged identical instances as different. I implemented proper hash there to make it hashable and usable in set comparision, and had to make a bit more complex comparision for dicts - to compare values by equality rather than identity.

* when serializing /deserializing we loose "DecoratedMappedOperator" - likely for a good reason, that's why I had to relax the check in __eq for type otherwise we would have DecoratedMappedOperator != MappedOperator. I think it might not matter that we are loosing Decorated (because I think they are equivalent after deserializing) but maybe it's worth looking at. 

* I excluded some fields from comparision, because they were wildly different after serialize/deserialize. I guess it does not matter as those are only used to before mapping  (serializing happens already for fully mapped task I think) but it woudl be great to double check if that is the right approach and whether we have not lost anything 

For dags:
      * 'partial_kwargs',
      * 'operator_class',
For tasks:
      * 'operator_class',
      * 'partial_kwargs',
      * 'mapped_op_kwargs',

**Pull request closed without merging**